src/js/components/List/List.js

```diff
@@ -277,16 +277,40 @@ const List = React.forwardRef(
                 let boxProps = {};
                 let itemId;
 
-                if (children) {
+                if (children && primaryKey) {
+                  content = children(
+                    item,
+                    index,
+                    onClickItem ? { active: active === index } : undefined,
+                  );
+                  itemId = normalize(item, index, primaryKey);
+                } else if (children) {
                   content = children(
                     item,
                     index,
                     onClickItem ? { active: active === index } : undefined,
                   );
                 } else if (primaryKey) {
                   if (typeof primaryKey === 'function') {
-                    itemId = primaryKey(item, index);
-                    content = itemId;
+                    content = primaryKey(item, index);
+                    if (!content?.key)
+                      console.error(
+                        `You must add key to your element in primaryKey func.`,
+                      );
+                    itemId = content.key;
+                  } else if (
+                    primaryKey.match(/^(.+),(.+)$/) &&
+                    primaryKey.split(',').length === 2
+                  ) {
```

Comment:
I'm still thinking about this section. I'm questioning if the pattern of having `primaryKey="displayValue, id"` is intuitive for users. We are deviating from the original purpose of primaryKey, I'm wondering if it would make more sense to have a separate prop for key instead of expanding the primaryKey prop to have multiple purposes and only use primaryKey as the key if a key value is not specifically given through the `key` prop.

Reply:
I agree with your points. Thank you for suggesting them. I will make the changes. What do you suggest should be the name of the prop ?  Maybe `listKey` ?

Reply:
I'm wondering if this is an infrequent enough scenario that we can just rely on the primaryKey as function approach?

Reply:
I’m thinking we should have a prop `itemKey`. I was leaning towards itemKey over listKey to convey to the user that the key is being applied to each item in the list not something that is being applied to the list as a whole. `itemKey` could be a string or a function.

Example data: 
const data = [
  { id: 1, location: 'Boise'},
  { id: 2, location: 'Fort Collins'},
  { id: 3, location: 'Los Gatos'},
  { id: 4, location: 'Palo Alto'},
  { id: 5, location: 'San Francisco'},
];

1. The user sets `itemKey` to a string `itemKey=“id”` 
The key should be equal to the id property for each item (ex 1, 2, 3…)
2. The `itemKey` prop is a function `itemKey={(item) => (key-${item.id})}` 
The key should be “key-<id property>” (ex. key-1, key-2, key-3…)
3. The user doesn’t set the `itemKey` prop but they do set `primaryKey`
 The `primaryKey` property should be used for the key (I believe this is how the key is currently determined)
4. The user doesn’t set `itemKey` or `primaryKey`
 We should default to the current behavior of using the index for the key (ex 0, 1, 2…)

@ericsoderberghp any thoughts on this direction?

Reply:
Oops I didn't see your comment above before I posted mine @ericsoderberghp

Reply:
I agree, I think we could just support setting the key within a `primaryKey` function and not support setting the key like this: `primaryKey="displayValue, id"`. Then we wouldn't need to add an additional prop.

Reply:
If we are going through the `primarKey` prop way, I thought this way `primaryKey="displayValue, id"`, works well, but if you guys have a better way, I am happy to implement that.

Reply:
I think we can drop support for `primaryKey="displayValue, id"` and just allow users to set the key using a `primaryKey` function ex. `primaryKey={(item) => (<Box key={item.id}>{item.location}</Box>)}`

Reply:
Hmm... I don't think we can use a React element as a `key` property: https://reactjs.org/docs/lists-and-keys.html#keys
Given that, and at the risk of being harangued with "I told you so" comments, I'm going back to @jcfilben 's excellent suggestion of adding an optional `itemKey` property that is optional and can be a string or a function as she outlined. Apologies.